File: sdk/ts-sdk/src/cgt.ts
/**
 * CGT (Creator God Token) operations
 */

import { DemiurgeClient } from './client';
import { CgtMetadata, CgtBalance, Address } from './types';
import { cgtFromSmallest, cgtToSmallest } from './utils';

export class CgtApi {
  constructor(private client: DemiurgeClient) {}

  /**
   * Get CGT metadata (name, symbol, decimals, max supply, total supply)
   */
  async getMetadata(): Promise<CgtMetadata> {
    return this.client.call<CgtMetadata>('cgt_getCgtMetadata');
  }

  /**
   * Get total supply of CGT
   */
  async getTotalSupply(): Promise<string> {
    const result = await this.client.call<{ totalSupply: string }>('cgt_getTotalSupply');
    return result.totalSupply;
  }

  /**
   * Get CGT balance for an address
   */
  async getBalance(address: Address): Promise<string> {
    const result = await this.client.call<CgtBalance>('cgt_getBalance', { address });
    return result.balance;
  }

  /**
   * Get CGT balance as a human-readable number
   */
  async getBalanceFormatted(address: Address): Promise<number> {
    const balance = await this.getBalance(address);
    return cgtFromSmallest(balance);
  }

  /**
   * Get transaction nonce for an address
   */
  async getNonce(address: Address): Promise<number> {
    return this.client.call<number>('cgt_getNonce', { address });
  }

  /**
   * Build a transfer transaction (does not sign or submit)
   */
  async buildTransfer(
    from: Address,
    to: Address,
    amount: number | string,
    nonce?: number
  ): Promise<{ unsignedTxHex: string; nonce: number }> {
    const amountStr = typeof amount === 'number' ? cgtToSmallest(amount) : amount;
    const txNonce = nonce ?? await this.getNonce(from);

    const result = await this.client.call<{ unsigned_tx_hex: string }>(
      'cgt_buildTransferTx',
      {
        from,
        to,
        amount: amountStr,
        nonce: txNonce,
      }
    );

    return {
      unsignedTxHex: result.unsigned_tx_hex,
      nonce: txNonce,
    };
  }

  /**
   * Submit a signed transaction
   */
  async sendRawTransaction(signedTxHex: string): Promise<string> {
    const result = await this.client.call<{ tx_hash: string }>(
      'cgt_sendRawTransaction',
      { tx_hex: signedTxHex }
    );
    return result.tx_hash;
  }

  /**
   * Get transaction by hash
   */
  async getTransaction(txHash: string): Promise<any> {
    return this.client.call('cgt_getTransaction', { hash: txHash });
  }

  /**
   * Get transaction history for an address
   */
  async getTransactionHistory(address: Address, limit?: number): Promise<any[]> {
    return this.client.call<any[]>('cgt_getTransactionHistory', {
      address,
      limit: limit ?? 100,
    });
  }

  /**
   * Request CGT from dev faucet (debug builds only)
   */
  async requestFaucet(address: Address): Promise<string> {
    const result = await this.client.call<{ tx_hash: string }>('cgt_devFaucet', {
      address,
    });
    return result.tx_hash;
  }
}

File: sdk/rust-sdk/src/cgt.rs
/*!
 * CGT (Creator God Token) operations
 */

use crate::client::DemiurgeClient;
use crate::error::Result;
use crate::types::{CgtBalance, CgtMetadata, ChainInfo};
use serde_json::json;

pub struct CgtApi<'a> {
    client: &'a DemiurgeClient,
}

impl<'a> CgtApi<'a> {
    pub fn new(client: &'a DemiurgeClient) -> Self {
        Self { client }
    }

    /// Get CGT metadata
    pub async fn get_metadata(&self) -> Result<CgtMetadata> {
        self.client.call("cgt_getCgtMetadata", None).await
    }

    /// Get total supply of CGT
    pub async fn get_total_supply(&self) -> Result<String> {
        #[derive(serde::Deserialize)]
        struct Response {
            total_supply: String,
        }
        let result: Response = self.client.call("cgt_getTotalSupply", None).await?;
        Ok(result.total_supply)
    }

    /// Get CGT balance for an address
    pub async fn get_balance(&self, address: &str) -> Result<String> {
        let result: CgtBalance = self
            .client
            .call("cgt_getBalance", Some(json!({ "address": address })))
            .await?;
        Ok(result.balance)
    }

    /// Get transaction nonce for an address
    pub async fn get_nonce(&self, address: &str) -> Result<u64> {
        self.client
            .call("cgt_getNonce", Some(json!({ "address": address })))
            .await
    }

    /// Get chain info
    pub async fn get_chain_info(&self) -> Result<ChainInfo> {
        self.client.call("cgt_getChainInfo", None).await
    }

    /// Submit a signed transaction
    pub async fn send_raw_transaction(&self, signed_tx_hex: &str) -> Result<String> {
        #[derive(serde::Deserialize)]
        struct Response {
            tx_hash: String,
        }
        let result: Response = self
            .client
            .call(
                "cgt_sendRawTransaction",
                Some(json!({ "tx_hex": signed_tx_hex })),
            )
            .await?;
        Ok(result.tx_hash)
    }

    /// Get transaction by hash
    pub async fn get_transaction(&self, tx_hash: &str) -> Result<serde_json::Value> {
        self.client
            .call("cgt_getTransaction", Some(json!({ "hash": tx_hash })))
            .await
    }
}

